alternator/auth.cc

```diff
@@ -51,7 +53,14 @@ future<std::string> get_key_from_roles(service::storage_proxy& proxy, auth::serv
     if (result_set->empty()) {
         co_await coroutine::return_exception(api_error::unrecognized_client(format("User not found: {}", username)));
     }
-    const managed_bytes_opt& salted_hash = result_set->rows().front().front(); // We only asked for 1 row and 1 column
+    const auto& result = result_set->rows().front();
+    bool can_login = result[1] && value_cast<bool>(boolean_type->deserialize(*result[1]));
+    if (!can_login) {
+        // This is a valid role name, but has "login=False" so should not be
+        // usable for authentication (see #19735).
+        co_await coroutine::return_exception(api_error::unrecognized_client(format("Role {} has login=false so cannot be used for login", username)));
+    }
```

Comment:
I wonder whether it's the right thing security-wise. Won't returning this specific error message allow unauthorized outsiders to learn whether a given non-login role exists or not? Is this error propagated back to the user in an unchanged form?

Given that descriptive errors are returned in other cases in this function as well, it looks like a pre-existing problem, so I don't think this is a blocker. Maybe the errors are appropriately censored already, or maybe learning whether a non-login role exists or not is not a big deal...

Reply:
I personally think that if this a problem at all, then maybe the problem is the pre-existing "User not found" error, not the new "login=false" error. What can attacker possibly do with the knowledge that a non-login role exists?

I'll add an issue to rethink all the error messages in this function. The important thing is the UnrecognizedClient error type - the error message doesn't need to give hints why this is the case.

Reply:
Opened https://github.com/scylladb/scylladb/issues/19852.